statsmodels/base/_screening.py

```diff
@@ -0,0 +1,323 @@
+                p[~keep] = 0
+            predicted = res_pen.model.predict(p)
+            # this is currently hardcoded for Poisson
+            resid_factor = (endog - predicted) / np.sqrt(predicted)
+        elif self.ranking_attr[:6] == 'model.':
+            # use model method, this is intended for score_factor
+            attr = self.ranking_attr.split('.')[1]
+            resid_factor = getattr(res_pen.model, attr)(res_pen.params)
+            if resid_factor.ndim == 2:
+                # for score_factor when extra params are in model
+                resid_factor = resid_factor[:, 0]
+            mom_cond = np.abs(resid_factor.dot(exog))**2
+        else:
+            # use results attribute
+            resid_factor = getattr(res_pen, self.ranking_attr)
+            mom_cond = np.abs(resid_factor.dot(exog))**2
+        return mom_cond
+
+    def screen_exog(self, exog, endog=None, maxiter=5, method='bfgs',
+                    disp=False):
+        """screen and select variables (columns) in exog
+
+        Parameters
+        ----------
+        exog : ndarray
+            candidate explanatory variables that are screened for inclusion in
+            the model
+        endog : ndarray (optional)
+            use a new endog in the screening model.
+            This is not tested yet, and might not work correctly
+        maxiter : int
+            number of screening iterations
+        method : str
+            optimization method to use in fit, needs to be only of the gradient
+            optimizers
+        disp : bool
+            display option for fit during optimization
+
+        Returns
+        -------
+        res_screen : instance of ScreeningResults
+            The attribute `results_final` contains is the results instance
+            with the final model selection.
+            `idx_nonzero` contains the index of the selected exog in the full
+            exog, combined exog that are always kept plust exog_candidates.
+
+
+        """
+        model_class = self.model_class
+        if endog is None:
+            # allow a different endog than used in model
+            endog = self.endog
+        x0 = self.exog_keep
+        x1 = exog
+        k0 = x0.shape[1]
+        # TODO: remove the need for x, use x1 separately from x0
+        # needs change to idx to be based on x1 (candidate variables)
+        x = np.column_stack((x0, x1))
+        nobs, k_vars = x.shape
+
+        history = defaultdict(list)
+        idx_nonzero = [0]
+        keep = np.array([True])
+        idx_excl = np.arange(1, k_vars)
+        #start_params = [endog.mean()]
+        res_pen = model_class(endog, x0, **self.init_kwds).fit(disp=disp)
+        start_params = res_pen.params
+        converged = False
+        idx_old = []
+        for it in range(maxiter):
+            mom_cond = self.ranking_measure(res_pen, x1, keep=keep)
+            mcs = np.sort(mom_cond)[::-1]
+
+            threshold = mcs[max((self.k_max_add, k0 + self.k_add))]
```

Comment:
maybe I wanted a `min` instead of `max` here
but because k0 is fixed, this would be constant

maybe k0 was supposed to be number of currently included exog.
Yes, in my original script version from 2015, I had current k_exog instead of k0 - refactoring mistake

Reply:
There are currently no checks that we don't run into boundary conditions with the indices, e.g. that we don't want to add more variables than available.

Also, if I change k0 to k_current, then it might be that k_current is k_max_add and we don't have space for additional variables anymore, i.e. adding set is empty and we have trivially converged (if we don't get an exception in this case)
Similarly if user specifies k_max_add < k0, then there is no space for screening additional exog.